File: bifrost-ts/src/server/event.ts
import { MessageType, Presentation } from '../utils'
import type { ClientNode } from './client-node'
import type { Server } from './server'
import type { ServerChannel } from './server-channel'

export type EventOptions = {
  /**
   * Requires authentication to subscribe to this event.
   */
  protected?: boolean

  /**
   * Only allow user to subscribe to this event on his own channel. Automatically makes the event protected.
   */
  user?: boolean

  /**
   * This overrides the `user` flag.
   */
  shouldSubscribe?: (
    client: ClientNode,
    eventName: string,
    channel: string,
  ) => Promise<boolean>

  /**
   * This will propagate the event to other instances when running node in a cluster.
   */
  cluster?: boolean

  /**
   * When true, excludes the originating client from receiving the event.
   * Requires `params.uuid` to identify the originator.
   */
  excludeOriginator?: boolean
}

export class Event {
  uuid: string
  name: string
  isProtected: boolean
  channel: ServerChannel
  server: Server
  cluster: boolean
  excludeOriginator: boolean

  shouldSubscribe: (
    client: ClientNode,
    eventName: string,
    channel: string,
  ) => Promise<boolean> = async () => true

  constructor(
    name: string,
    server: Server,
    channel: ServerChannel,
    opts?: EventOptions,
  ) {
    this.uuid = Presentation.uuid()
    this.name = name
    this.server = server
    this.channel = channel

    this.isProtected = opts?.protected ?? false

    if (opts?.user) {
      this.isProtected = true

      this.shouldSubscribe = async function (client, event, channel) {
        if (!client.userId) return false

        return channel === client.userId.toString()
      }
    }

    if (opts?.shouldSubscribe) {
      this.shouldSubscribe = opts.shouldSubscribe
    }

    this.cluster = Boolean(opts?.cluster)
    this.excludeOriginator = Boolean(opts?.excludeOriginator)
  }

  /**
   * Encodes the event with the wire protocol `t` field so propagate()
   * can forward the payload directly without re-encoding.
   */
  handler(channel: ServerChannel, params: Record<string, string>): void {
    const payload = Presentation.encode({
      t: MessageType.EVENT,
      uuid: Presentation.uuid(),
      event: this.name,
      channel: channel.channelName,
      params,
    })

    const excludeUuid = this.excludeOriginator ? params.uuid : undefined

    if (this.cluster && this.server?.redisTransport?.pub) {
      this.server.redisTransport
        .publish(this.name, channel.channelName, payload, excludeUuid)
        .catch(console.error)

      return
    }

    channel.propagate(this.name, payload, excludeUuid)
  }
}

File: typeferry-py/src/typeferry/server/event.py
"""Event primitive — port of ``typeferry-ts/src/server/event.ts``.

An :class:`Event` represents a declared subscription channel topic.
When emitted, it encodes the wire envelope (``t: "event"``) and routes
either to the local :class:`RoomRegistry` or through Redis for
multi-instance deployments.
"""

from __future__ import annotations

import inspect
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

from typeferry.ejson.presentation import Presentation
from typeferry.protocol.messages import MessageType

if TYPE_CHECKING:
    from typeferry.server.client_node import ClientNode
    from typeferry.server.server import Server
    from typeferry.server.server_channel import ServerChannel


ShouldSubscribe = Callable[
    ["ClientNode", str, str], bool | Awaitable[bool]
]


@dataclass(slots=True)
class EventOptions:
    protected: bool = False
    """Require authentication to subscribe."""

    user: bool = False
    """Implies ``protected``; only allow a user to subscribe on their own channel."""

    should_subscribe: ShouldSubscribe | None = None
    """Custom async predicate gating subscription. Overrides ``user``."""

    cluster: bool = False
    """Propagate via Redis to other instances."""

    exclude_originator: bool = False
    """When True, uses ``params['uuid']`` to identify and skip the originator."""


async def _always_true(_client: Any, _event: str, _channel: str) -> bool:
    return True


class Event:
    """Declared event with subscription policy.

    Subscription policy is evaluated in this order (matches TS):

    1. explicit ``should_subscribe`` if provided (wins over ``user``)
    2. ``user`` flag — channel MUST equal ``client.user_id``
    3. default → always allow (authorization handled by ``is_protected``)
    """

    uuid: str
    name: str
    server: Server
    channel: ServerChannel
    is_protected: bool
    cluster: bool
    exclude_originator: bool
    should_subscribe: ShouldSubscribe

    def __init__(
        self,
        name: str,
        server: Server,
        channel: ServerChannel,
        opts: EventOptions | None = None,
    ) -> None:
        opts = opts or EventOptions()
        self.uuid = Presentation.uuid()
        self.name = name
        self.server = server
        self.channel = channel
        self.is_protected = opts.protected

        self.should_subscribe = _always_true

        if opts.user:
            self.is_protected = True
            self.should_subscribe = _user_scoped_subscribe

        if opts.should_subscribe is not None:
            self.should_subscribe = opts.should_subscribe

        self.cluster = bool(opts.cluster)
        self.exclude_originator = bool(opts.exclude_originator)

    async def handler(
        self, channel: ServerChannel, params: Any
    ) -> Awaitable[None] | None:
        """Encode the event and dispatch locally or via Redis."""

        payload = Presentation.encode(
            {
                "t": MessageType.EVENT.value,
                "uuid": Presentation.uuid(),
                "event": self.name,
                "channel": channel.channel_name,
                "params": params,
            }
        )

        exclude_uuid: str | None = None
        if self.exclude_originator and isinstance(params, dict):
            raw = params.get("uuid")
            if isinstance(raw, str):
                exclude_uuid = raw

        if self.cluster and self.server.redis_transport is not None:
            await self.server.redis_transport.publish(
                self.name, channel.channel_name, payload, exclude_uuid
            )
            return None

        await channel.propagate(self.name, payload, exclude_uuid)
        return None

    async def evaluate_should_subscribe(
        self, client: ClientNode, event: str, channel: str
    ) -> bool:
        result = self.should_subscribe(client, event, channel)
        awaited = await result if inspect.isawaitable(result) else result
        return bool(awaited)


async def _user_scoped_subscribe(
    client: ClientNode, _event: str, channel: str
) -> bool:
    if not client.user_id:
        return False
    return str(client.user_id) == channel
